more for fix #157: update detect vs2017 scripts to v1.11.0

refack opened the pull request.

KindDragon commented:
Hi. I reworked your script in PR #173 to support checking `Microsoft.VisualCpp.Tools.Core` version and return only string instead of JSON

KindDragon commented on src/tools/vc141helper/GetKey.ps1:
```diff
@@ -0,0 +1,16 @@
+    [string]$key = "VisualCppToolsCL"
+)
+Add-Type -Path GetVS2017Configuration.cs;
+$inst = ([VisualStudioConfiguration.ComSurrogate]::QueryEx($filter))[0]
```
It can be empty list

refack replied:
Changed

refack commented:
@KindDragon I also refactored the JSON code to a different file, added "IsVcCompatible" predicate, that checks for the minimal VC triad: `"SDK" && "MSBuild" && "VisualCppTools"`, and streamlined the `.ps1`

refack commented:
@KindDragon P.S. how did you get github to acknowledge AppYevor?

KindDragon commented:
> @KindDragon I also refactored the JSON code to a different file, added "IsVcCompatible" predicate, that checks for the minimal VC triad: "SDK" && "MSBuild" && "VisualCppTools", and streamlined the .ps1

But you not checking visual studio version

> @KindDragon P.S. how did you get github to acknowledge AppYevor?

AppVeyor build every PR for this repository https://ci.appveyor.com/project/boostorg/build/build/1.0.106

BTW because of PowerShell test run 33 minutes for VS2017 and only 6 min for other versions. I will try switching to `vswhere` to speed up tests

refack commented:
> But you not checking visual studio version

What do you mean? You meen future proofing (since for now it's only 15.0)?

> BTW because of PowerShell tests run 33 minutes for VS2017 and only 6 min for other versions. I will try switching to vswhere to speed up tests

Let me look into that... (since the actual `bootstrap.bat` and `bjam` work reasonably fast...)

KindDragon commented:
> What do you mean? You meen future proofing (since for now it's only 15.0)?

It will probably work for next VS after 2017

> Let me look into that... (since the actual bootstrap.bat and bjam work reasonably fast...)

I see in Process Explorer spawning a lot of PowerShell executions. I think each test run your script many times

refack commented:
> I see in Process Explorer spawning a lot of PowerShell executions. I think each test run your script many times

That's a bug i've seen in `bootstrap.bat` when something fails...

refack commented:
I.E. look at how it fails https://ci.appveyor.com/project/boostorg/build/build/1.0.106/job/1fabtlet1fdlmr6o
`******  B A T C H   R E C U R S I O N  exceeds STACK limits ******`

KindDragon commented:
> I.E. look at how it fails https://ci.appveyor.com/project/boostorg/build/build/1.0.106/job/1fabtlet1fdlmr6o
> ****** B A T C H R E C U R S I O N exceeds STACK limits ******

I fixed this in my PR by splitting build.bat to several files. I tried many options, but only it helped

refack commented:
Yea I can relate. `cmd` is super fickle 🤦

refack commented:
Green on [my fork](https://ci.appveyor.com/project/refack/build/build/1.0.8) in good time

@grafikrobot you can cancel all others builds

refack commented:
@grafikrobot would you consider adding the COM helper tools as a [git submodule](https://github.com/node4good/msvs-com-helper)?

KindDragon commented on test/test_all.py:
```diff
@@ -284,6 +283,7 @@ def reorder_tests(tests, first_test):
 
 if os.name == "posix":
     tests.append("symlink")
+    tests.append("link")
```
This test works on Windows for me, but not on AppVeyor

refack replied:
Fails for me locally as well. I opened #185, will look into it soon.

KindDragon commented on src/engine/guess_toolset.bat:
```diff
@@ -32,11 +33,24 @@ if NOT "_%VS150COMNTOOLS%_" == "__" (
+:vsCOMhelper_2
+IF "_%~1_"=="__" goto :eof
+set "BOOST_JAM_TOOLSET_ROOT=%~1\VC\"
+for /f "tokens=*" %%A in ('cmd /D /S /C "%vsCOMhelper_cmd% VCToolsVersionCode"') do if NOT "_%%A_"=="__" (
```
Who will check version from `VCToolsVersionCode`?

refack replied:
So it's forward compatible. Bootstrap will fail with `Unknown toolset: %BOOST_JAM_TOOLSET%`

KindDragon replied:
I just can't find where get_key_helper.cmd checking `VCToolsVersionCode`

refack replied:
The full command will be `"%~dp0..\tools\vsCOMhelper\get_key_helper.cmd" IsVcCompatible VCToolsVersionCode`
first arg `IsVcCompatible` is [a filter ](https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R15)
second arg `VCToolsVersionCode` is[ a projection](https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R27)

KindDragon commented on src/engine/config_toolset.bat:
```diff
@@ -165,21 +166,25 @@ if "_%BOOST_JAM_TOOLSET_ROOT%_" == "__" (
+
+IF DEFINED EXPENSIVE_POWERSHELL2 goto :Skip_VC1410
+SET EXPENSIVE_POWERSHELL2=1
+SET vsCOMhelper_cmd="%~dp0..\tools\vsCOMhelper\get_key_helper.cmd" "IsVcCompatible,VisualCppToolsVersionMinor=14.10" "InstallationPath"
```
I can't find where Powershell/CSharp script check this version

refack replied:
https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R17

refack replied:
My concept is; `C#` only does marshaling of data from COM, `powershell` does filtering / projecting.

refack replied:
@KindDragon I took the filtering idea from you, it's a very good idea 👍 I'm using it in all other projects...

KindDragon commented on src/engine/guess_toolset.bat:
```diff
@@ -1,4 +1,5 @@
-@ECHO OFF
+@IF NOT DEFINED DEBUG_BATCH @ECHO OFF
```
I just comment `@echo off` if I need debugging 😄

refack replied:
It goes well with https://github.com/boostorg/build/pull/179/files#diff-180360612c6b8c4ed830919bbb4dd459R14

refack replied:
BTW: stole the concept from VS2017 `VsDevCmd.bat` line 3: `@if NOT "%VSCMD_DEBUG%" GEQ "3" @echo off`

KindDragon commented on appveyor.yml:
```diff
@@ -11,27 +11,30 @@ image:
-    }
-    if ($TestExitCode -ne 0) { $host.SetShouldExit($TestExitCode) }
+    
+    python test_all.py msvc $env:TEST_ALL_EXTRAS --preserve
```
It should begin with &

refack replied:
Not sure why? This works (as it should).

P.S. __&__ operator does not parse variables.
From the docs:

__&__ Call operator
==
Run a command, script, or script block.
The call operator, also known as the "invocation operator," lets you run commands that are stored in variables and represented by strings.
__Because the call operator does not parse the command, it cannot interpret command parameters.__

```
C:\PS> $c = "get-executionpolicy"
C:\PS> $c
get-executionpolicy
C:\PS> & $c
AllSigned
```

KindDragon replied:
You right

refack replied:
http://stackoverflow.com/questions/42382951/whats-the-command-of-call-operator
Learned something new: You need the __&__ if you want to run a command that's not in enf:Path with a space in it's name, i.e.:
```
& "C:\bin\dev\VS\SDKs\10\bin\x64\rc.exe" $param1 $param2 $param3
```

refack replied:
I already know you can use it to run `{ script blocks }`
```
$reusable = { msbuild "node.sln" /verylong:parameter1 /verylong:parameter2 } //This one doesn't run
& $reusable 'nobuild';
& $reusable 'build';
& $reusable 'test';

refack commented:
Made obsolete by official install of `vswhere`

AndrewPardoe commented:
The officially supported vswhere tool is now available as part of the install starting today with Visual Studio 15.2 preview 2: https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/. This doesn't fix all the existing installs of VS 2017 but going forward you can now detect the VS install locations and whether C++ tools are available from a scripted environment. I hope this helps your scenario.
